File: scripts/translate_pw_gen.js
#!/usr/bin/env node
// Translate password-generator from CN → EN
const fs = require('fs');

let cn = fs.readFileSync('password-generator/index.html', 'utf8');

// Key content area translations (targeted to specific elements)
const contentReplacements = [
  // Hero area
  ['<p>免费在线密码生成器，支持<strong>常规随机密码</strong>、<strong>易记口令短语</strong>和<strong>PIN码</strong>三种模式。自定义字符集、排除相似字符、批量生成、多格式导出、📱扫描二维码传输到手机。所有密码使用', '<p>Free online password generator with <strong>random passwords</strong>, <strong>memorable passphrases</strong>, and <strong>PIN codes</strong>. Custom charsets, exclude similar chars, batch generation, multi-format export, 📱 QR code sharing to phone. All passwords use'],
  ['真随机生成，纯本地处理，绝不上传。</p>', 'true random generation, processed locally, never uploaded.</p>'],
  ['⚡ 实时生成 · 🔒 零泄露 · 📦 批量导出 · 💬 口令短语 · 📱 二维码分享 · 🧠 记忆评分', '⚡ Real-time · 🔒 Zero-leak · 📦 Batch Export · 💬 Passphrase · 📱 QR Share · 🧠 Memorability'],
  // Mode tabs
  ['🔀 随机密码', '🔀 Random'],
  ['💬 口令短语', '💬 Passphrase'],
  ['📱 PIN码', '📱 PIN'],
  // Settings headers
  ['⚙️ 随机密码选项', '⚙️ Random Password Options'],
  ['💬 口令短语选项', '💬 Passphrase Options'],
  ['📱 PIN码选项', '📱 PIN Options'],
  // Labels
  ['密码长度', 'Password Length'],
  ['生成数量', 'Count'],
  ['包含字符类型', 'Character Types'],
  ['大写字母 (A-Z)', 'Uppercase (A-Z)'],
  ['小写字母 (a-z)', 'Lowercase (a-z)'],
  ['数字 (0-9)', 'Numbers (0-9)'],
  ['特殊符号 (!@#$%^&*)', 'Symbols (!@#$%^&*)'],
  ['排除相似字符 (0 O 1 l I)', 'Exclude Similar (0 O 1 l I)'],
  ['自定义字符集', 'Custom Charset'],
  ['覆盖上方选择', 'overrides above'],
  ['留空使用上方字符类型，如：0123456789abcdef', 'Leave empty to use above, e.g.: 0123456789abcdef'],
  ['单词数量', 'Word Count'],
  ['分隔符', 'Separator'],
  ['- 连字符', '- Hyphen'],
  ['空格', 'Space'],
  ['. 点', '. Dot'],
  ['_ 下划线', '_ Underscore'],
  ['无分隔', 'No Sep'],
  ['首字母大写', 'Capitalize'],
  ['末尾添加数字', 'Add Number'],
  ['PIN长度', 'PIN Length'],
  ['PIN格式', 'PIN Format'],
  ['无格式', 'Plain'],
  ['每4位空格', '4-digit Space'],
  ['每4位连字符', '4-digit Dash'],
  // Buttons
  ['🔄 重新生成', '🔄 Regenerate'],
  ['📋 复制全部', '📋 Copy All'],
  ['💾 导出', '💾 Export'],
  ['导出格式：', 'Export Format: '],
  ['🗑️ 清除历史', '🗑️ Clear History'],
  // Results area
  ['📋 生成的密码', '📋 Generated Passwords'],
  ['自动复制', 'Auto Copy'],
  ['每次生成后自动复制第一个密码到剪贴板', 'Auto-copy first password on generation'],
  ['🕒 复制历史', '🕒 Copy History'],
  ['保存在本地浏览器中（localStorage），最多10条。数据不会上传到服务器。', 'Stored in local browser (localStorage), max 10 entries. Data is never uploaded.'],
  // Tutorial
  ['📖 使用教程', '📖 How to Use'],
  ['模式一：随机密码', 'Mode 1: Random Password'],
  ['模式二：口令短语', 'Mode 2: Passphrase'],
  ['模式三：PIN码', 'Mode 3: PIN Code'],
  ['一键导出', 'One-click Export'],
  ['键盘快捷键', 'Keyboard Shortcuts'],
  ['🔍 密码泄露检测', '🔍 Leak Detection'],
  ['📋 自动复制模式', '📋 Auto Copy Mode'],
  // Use Cases
  ['🎯 应用场景', '🎯 Use Cases'],
  ['💡 安全密码建议', '💡 Password Security Tips'],
  // FAQ
  ['❓ 常见问题', '❓ FAQ'],
  ['🔒 密码强度说明', '🔒 Password Strength Guide'],
  ['弱', 'Weak'],
  ['强', 'Strong'],
  ['非常强', 'Very Strong'],
  ['⚡ 快捷预设', '⚡ Quick Presets'],
  ['⌨️ 快捷键', '⌨️ Shortcuts'],
  ['🆕 新功能', '🆕 New Features'],
  ['🧠 记忆难度评估', '🧠 Memorability Assessment'],
  ['🔗 相关工具推荐', '🔗 Related Tools'],
];

// Apply content replacements
for (const [from, to] of contentReplacements) {
  cn = cn.split(from).join(to);
}

// Global replacements for common terms appearing in many places  
const globalReplacements = [
  ['在线密码生成器 · 纯前端本地生成 · 密码绝不上传服务器', 'Online Password Generator · Pure Frontend · Passwords Never Uploaded'],
  ['问题反馈: dexshuang@google.com', 'Feedback: dexshuang@google.com'],
  ['选择密码长度（4-128位）和字符类型，勾选"排除相似字符"可去掉0/O/1/l/I。滑块拖动时', 'Choose length (4-128) and character types. Check "Exclude Similar" to remove 0/O/1/l/I. Sliders update '],
  ['实时生成', 'in real-time'],
  ['，你也可以使用', '. You can also use '],
  ['（如Hex字符集：0123456789abcdef）来生成满足特定要求的密码。', ' (e.g., Hex: 0123456789abcdef) for specific requirements.'],
  ['由3-10个随机英文单词组成，如', '3-10 random English words, e.g. '],
  ['。可选分隔符（连字符/空格/点/下划线）、首字母大写、末尾添加数字。口令短语比随机字符串更易记忆，同时具有极高安全性。', '. Optional separator, capitalization, trailing number. Passphrases are easier to remember while maintaining high security.'],
  ['纯数字PIN码，4-16位。支持无格式、四位空格分组、四位连字符分组三种格式。适合银行PIN、安全码、验证码等场景。', 'Numeric PIN, 4-16 digits. Plain, 4-digit space, or 4-digit dash formats. Ideal for bank PINs, security codes, verification codes.'],
  ['点击导出按钮，支持TXT、CSV、JSON三种格式下载。适合批量保存或导入密码管理器。', 'Click export for TXT, CSV, JSON download. Great for batch saving or importing to password managers.'],
  ['重新生成', 'Regenerate'],
  ['复制第一个密码', 'Copy First'],
  ['点击第一个密码旁的🔍按钮，即可通过Have I Been Pwned的k-anonymity API检查该密码是否出现在已知数据泄露中。完整密码不会被发送——仅哈希前5位用于查询，绝对安全。', 'Click the 🔍 button next to the first password to check via Have I Been Pwned k-anonymity API. Only the first 5 chars of the SHA-1 hash are sent — completely safe.'],
  ['勾选"自动复制"复选框后，每次生成密码时第一个密码会自动复制到剪贴板，省去手动点击步骤。', 'Check "Auto Copy" to automatically copy the first password on each generation.'],
  // 404 footer
  ['首页', 'Home'],
  ['工具', 'Tools'],
  ['全部工具', 'All Tools'],
  ['联系我们', 'Contact'],
  ['隐私政策', 'Privacy'],
  ['服务条款', 'Terms'],
  ['关于我们', 'About'],
  // Nav back
  ['首页</a>', 'Home</a>'],
  // Toast messages
  ['已复制密码', 'Password copied'],
  ['已复制', 'Copied'],
  ['已复制全部', 'Copied all'],
  ['已复制全部短语', 'Copied all phrases'],
  ['已复制全部PIN码', 'Copied all PINs'],
  ['请先生成密码', 'Generate passwords first'],
  ['请先生成口令短语', 'Generate passphrases first'],
  ['请先生成PIN码', 'Generate PINs first'],
  ['复制失败', 'Copy failed'],
  ['已清除历史记录', 'History cleared'],
  ['🔒 已自动复制密码到剪贴板', '🔒 Auto-copied to clipboard'],
  ['📱 扫描二维码复制到手机', '📱 Scan QR to copy to phone'],
  ['用手机相机扫描二维码，密码将自动复制', 'Scan with phone camera to transfer password'],
  ['关闭', 'Close'],
  ['⚠️ 此密码曾出现在泄露数据库中，建议更换！', '⚠️ This password appeared in known data breaches — change it!'],
  ['✅ 未在已知泄露中发现此密码', '✅ Password not found in known breaches'],
  ['泄露检测失败，请检查网络后重试', 'Leak check failed, check network and retry'],
  ['预估破解时间', 'Est. crack time'],
  ['检测泄露', 'Check Leak'],
  ['生成二维码 - 手机扫描复制', 'Generate QR - scan with phone'],
  ['已复制第一个密码', 'Copied first password'],
  ['重新复制', 'Re-copy'],
  ['复制', 'Copy'],
  ['请选择字符类型并生成密码', 'Select character types and generate'],
  ['请至少选择一种字符类型', 'Please select at least one character type'],
  ['自定义字符集至少需要2个字符', 'Custom charset needs at least 2 characters'],
  ['没有可导出的密码', 'No passwords to export'],
  ['已导出', 'Exported'],
  ['个密码', ' passwords'],
  ['平均熵', 'Avg entropy'],
  ['位', ' bits'],
  ['最弱密码破解', 'Weakest crack'],
  ['个易记', ' easy'],
  ['个中等', ' medium'],
  ['难记', 'Hard'],
  ['易记', 'Easy'],
  ['中等', 'Medium'],
  ['评分', ''],
  ['每个密码附带', 'Each password rated: '],
  ['综合考虑长度、字符类型、模式', 'Considers length, char types, and pattern'],
  ['📱 二维码分享 (手机扫描复制)', '📱 QR Share (phone scan)'],
  ['🧠 记忆难度评分', '🧠 Memorability Score'],
  ['🔍 密码泄露检测 (HIBP)', '🔍 Leak Check (HIBP)'],
  ['⏱ 破解时间估算', '⏱ Crack Time Estimate'],
  ['📋 自动复制模式', '📋 Auto Copy Mode'],
  ['💾 历史跨会话保存', '💾 Cross-session History'],
  ['📚 500+口令词库', '📚 500+ Word Library'],
  ['🔐 Hash生成器', '🔐 Hash Generator'],
  ['📝 Base64编解码', '📝 Base64 Encoder'],
  ['🔐 AES加密', '🔐 AES Encrypt'],
  ['🆔 UUID生成器', '🆔 UUID Generator'],
  ['64位十六进制密钥', '64-char hex key'],
  ['32字符Base32编码', '32-char Base32'],
  ['16位无符号', '16-char no symbols'],
  ['6-16位数字串', '6-16 digit string'],
  ['瞬间', 'Instant'],
  ['秒', 's'],
  ['分钟', 'min'],
  ['小时', 'hrs'],
  ['天', 'days'],
  ['年', 'years'],
  ['世纪', 'centuries'],
  ['万年', '10k years'],
  ['亿年', '100M years'],
  ['未在已知泄露中发现此密码', 'Not found in known breaches'],
  ['此密码曾在数据泄露中出现过！建议更换', 'Found in data breaches! Change recommended'],
  ['请选择字符类型并生成密码', 'Select character types and generate'],
  ['生成', 'Generate'],
];

for (const [from, to] of globalReplacements) {
  cn = cn.split(from).join(to);
}

// Fix lang attribute
cn = cn.replace('lang="zh-CN"', 'lang="en"');
cn = cn.replace(/hreflang="zh"/g, 'hreflang="en"');

// Fix title
cn = cn.replace(/<title>.*?<\/title>/, '<title>Free Password Generator - Strong Random Passwords | Passphrase | PIN | Real-time</title>');

// Fix OG URLs
cn = cn.replace(/content="https:\/\/free-toolbase\.com\/password-generator\/"/g, 'content="https://free-toolbase.com/en/password-generator/"');
cn = cn.replace(/content="https:\/\/free-toolbase\.com\/"/g, 'content="https://free-toolbase.com/en/"');
cn = cn.replace(/canonical" href="https:\/\/free-toolbase\.com\/password-generator\/"/, 'canonical" href="https://free-toolbase.com/en/password-generator/"');

// Fix OG image
cn = cn.replace(/content="https:\/\/free-toolbase\.com\/og-image\.svg"/g, 'content="https://free-toolbase.com/og-image.svg"');
cn = cn.replace(/og-image\.svg/g, '../../og-image.svg');

// Fix language switch
cn = cn.replace(
  '<a href="index.html" class="active">中文</a><a href="../en/password-generator/">EN</a>',
  '<a href="../../password-generator/">中文</a><a href="index.html" class="active">EN</a>'
);

// Fix nav links
cn = cn.replace(/href="\.\.\/index\.html"/g, 'href="../index.html"');
cn = cn.replace(/href="\.\.\/privacy\/"/g, 'href="../../privacy/"');
cn = cn.replace(/href="\.\.\/terms\/"/g, 'href="../../terms/"');
cn = cn.replace(/href="\.\.\/about\/"/g, 'href="../../about/"');

// Fix tool links to en versions
cn = cn.replace(/href="\/hash-generator\//g, 'href="/en/hash-generator/"');
cn = cn.replace(/href="\/base64\//g, 'href="/en/base64/"');
cn = cn.replace(/href="\/aes-encrypt\//g, 'href="/en/aes-encrypt/"');
cn = cn.replace(/href="\/uuid-generator\//g, 'href="/en/uuid-generator/"');

// Fix footer link back to EN homepage  
cn = cn.replace(/href="\.\.\/index\.html"/g, 'href="../index.html"');
cn = cn.replace(/href="\.\.\/\.\.\/index\.html"/g, 'href="../index.html"');
cn = cn.replace(/href="\.\.\/\.\.\/privacy\//g, 'href="../../privacy/"');
cn = cn.replace(/href="\.\.\/\.\.\/terms\//g, 'href="../../terms/"');
cn = cn.replace(/href="\.\.\/\.\.\/about\//g, 'href="../../about/"');

// Fix Schema description
cn = cn.replace(/"description":"免费在线密码生成器[^"]*"/, '"description":"Free online password generator with crypto.getRandomValues true random algorithm. Random passwords, passphrases, PIN codes, custom charset, exclude similar chars, batch export TXT/CSV/JSON, QR code sharing, memorability scoring, leak detection. Pure frontend, zero upload."');
cn = cn.replace(/"name":"在线密码生成器"/g, '"name":"Password Generator"');

// Fix remaining "位" in context like "240 位"
cn = cn.replace(/([\d.]+) 位/g, '$1 bits');
cn = cn.replace(/([\d.]+)位/g, '$1 bits');

// Fix the Japanese text remnants - "个密码" 
cn = cn.replace(/个/g, '');

// Fix footer link to EN in footer (already has ../../en/... from CN copy)
cn = cn.replace(/href="\.\.\/en\/password-generator\/"/g, 'href="../password-generator/"');

fs.writeFileSync('en/password-generator/index.html', cn);
console.log('English version created. Lines:', cn.split('\n').length);

File: scripts/translate_pwgen.py
#!/usr/bin/env python3
"""Translate password-generator CN -> EN"""
import re

with open('password-generator/index.html') as f:
    cn = f.read()

trans_map = {
    '密码生成器 - Free ToolBase': 'Password Generator - Free ToolBase',
    '免费在线密码生成器，使用crypto.getRandomValues生成真正随机的密码。支持随机密码、口令短语、PIN码三种模式。': 'Free online password generator using crypto.getRandomValues for truly random passwords. Random, passphrase & PIN code modes.',
    '密码生成器,随机密码,强密码,口令短语,PIN码,批量导出,在线工具,免费,安全密码': 'password generator, random password, strong password, passphrase, PIN code, batch export, online tool, free, secure password',
    '密码生成器': 'Password Generator',
    '生成真正随机的强密码，纯前端处理，密码不会离开你的浏览器': 'Generate truly random strong passwords. Pure frontend — passwords never leave your browser.',
    '随机密码': 'Random Password',
    '口令短语': 'Passphrase',
    'PIN码': 'PIN Code',
    '密码长度': 'Password Length',
    '生成数量': 'Count',
    '字符类型': 'Character Types',
    '大写字母': 'Uppercase',
    '小写字母': 'Lowercase',
    '数字': 'Numbers',
    '符号': 'Symbols',
    '排除相似字符': 'Exclude Similar',
    '自定义字符集': 'Custom Charset',
    '自定义': 'Custom',
    '快捷预设': 'Quick Presets',
    '重新生成': 'Regenerate',
    '复制全部': 'Copy All',
    '导出': 'Export',
    '生成的密码': 'Generated Passwords',
    '自动复制': 'Auto Copy',
    '清除历史': 'Clear History',
    '复制历史': 'Copy History',
    '保存在本地浏览器中': 'Stored in your browser',
    '数据不会上传到服务器': 'Data never leaves your device',
    '使用教程': 'How to Use',
    '模式一：': 'Mode 1: ',
    '模式二：': 'Mode 2: ',
    '模式三：': 'Mode 3: ',
    '选择密码长度': 'Select password length',
    '和字符类型': 'and character types',
    '勾选"排除相似字符"可去掉0/O/1/l/I': 'Check "Exclude Similar" to remove 0/O/1/l/I',
    '滑块拖动时': 'Drag the slider to ',
    '实时生成': 'generate in real-time',
    '你也可以使用': 'You can also use the',
    '来生成满足特定要求的密码': 'for specific requirements',
    '由3-10个随机英文单词组成': '3-10 random English words',
    '可选分隔符': 'Optional separator',
    '连字符': 'hyphen',
    '空格': 'space',
    '点': 'dot',
    '下划线': 'underscore',
    '无分隔': 'no separator',
    '首字母大写': 'Capitalize',
    '末尾添加数字': 'Add number',
    '口令短语比随机字符串更易记忆，同时具有极高安全性': 'Passphrases are easier to remember than random strings while remaining highly secure.',
    '纯数字PIN码': 'Pure numeric PIN codes',
    '支持无格式、四位空格分组、四位连字符分组三种格式': 'Plain, space-grouped, and dash-grouped formats.',
    '适合银行PIN、安全码、验证码等场景': 'Ideal for bank PINs, security codes, and verification codes.',
    '一键导出': 'One-click Export',
    '点击导出按钮，支持TXT、CSV、JSON三种格式下载。适合批量保存或导入密码管理器。': 'Export to TXT, CSV, or JSON. Great for batch saving or importing to password managers.',
    '键盘快捷键': 'Keyboard Shortcuts',
    '复制第一个密码': 'Copy First',
    '密码泄露检测': 'Leak Detection',
    '点击第一个密码旁的': 'Click ',
    '按钮，即可通过Have I Been Pwned的k-anonymity API检查该密码是否出现在已知数据泄露中。完整密码不会被发送——仅哈希前5位用于查询，绝对安全。': ' to check via Have I Been Pwned k-anonymity API. Only the first 5 chars of the hash are sent — your full password stays private.',
    '自动复制模式': 'Auto Copy Mode',
    '勾选"自动复制"复选框后，每次生成密码时第一个密码会自动复制到剪贴板，省去手动点击步骤。': 'When enabled, the first password is automatically copied to your clipboard on each generation.',
    '应用场景': 'Use Cases',
    '账号注册：': 'Account Registration: ',
    '为各大网站和应用生成高强度唯一密码': 'Generate strong unique passwords for websites and apps',
    '密码管理：': 'Password Management: ',
    '配合密码管理器使用，定期更换重要账号密码': 'Use with password managers, rotate important credentials regularly',
    'WiFi密码：': 'WiFi Passwords: ',
    '生成复杂度高的无线网络密码防止蹭网': 'Generate complex WiFi passwords to prevent piggybacking',
    'API密钥：': 'API Keys: ',
    '为开发项目生成Hex/Bas32格式的安全密钥': 'Generate secure Hex/Base32 keys for development projects',
    '银行PIN：': 'Bank PINs: ',
    '生成安全的4-6位数字PIN码': 'Generate secure 4-6 digit PIN codes',
    '易记密码：': 'Memorable Passwords: ',
    '使用口令短语模式创建既安全又好记的密码': 'Create secure yet memorable passwords with passphrase mode',
    '安全密码建议': 'Password Security Tips',
    '长度优先：': 'Length First: ',
    '密码长度比复杂度更重要。建议使用至少12位以上的密码，16位为佳。': 'Password length matters more than complexity. Use at least 12 characters, 16 is better.',
    '口令短语：': 'Passphrase: ',
    '5个随机单词的口令短语比12位随机字符密码更安全且更容易记忆。熵值可达100+位。': 'A 5-word passphrase is more secure and easier to remember than a 12-char random password. 100+ bits of entropy.',
    '混合字符：': 'Mix Characters: ',
    '包含大写字母、小写字母、数字和特殊符号的组合，大幅增加暴力破解难度。': 'Combine uppercase, lowercase, numbers, and symbols to dramatically increase brute-force resistance.',
    '避免规律：': 'Avoid Patterns: ',
    '不要使用有意义的单词、生日、电话号码等容易被猜测的信息。': 'Don\'t use dictionary words, birthdays, phone numbers, or other guessable information.',
    '唯一性：': 'Uniqueness: ',
    '不同网站使用不同密码，避免一个密码泄露导致所有账号被盗。': 'Use different passwords for each site — one breach shouldn\'t compromise everything.',
    '定期更换：': 'Rotate Regularly: ',
    '重要账号每3-6个月更换一次密码，并启用双因素认证(2FA)。': 'Rotate important passwords every 3-6 months and enable two-factor authentication (2FA).',
    '常见问题': 'FAQ',
    '生成的密码真的安全吗？': 'Are the generated passwords really secure?',
    '是的。本工具使用浏览器的 crypto.getRandomValues() 方法生成真正的随机密码，所有计算在您本地设备的浏览器中完成，生成的密码不会通过网络传输或存储在服务器上。': 'Yes. This tool uses the browser\'s crypto.getRandomValues() for true randomness. All computation happens locally — passwords are never transmitted or stored on any server.',
    '密码泄露检测是什么？': 'What is password leak detection?',
    '点击密码旁的': 'Click ',
    '按钮，工具会将密码的SHA-1哈希前5位发送到Have I Been Pwned的k-anonymity API，检查该密码是否出现在已知的数据泄露中。整个过程您的完整密码不会被传输到任何服务器，只有哈希的前5位被用于查询。': ' to check against Have I Been Pwned via k-anonymity API. Only the first 5 SHA-1 hash characters are sent for querying — your full password never leaves your browser.',
    '破解时间是如何计算的？': 'How is crack time calculated?',
    '基于密码熵值（信息熵）和现代GPU集群每秒1000亿次猜测的能力进行估算。熵值越高，破解所需时间越长。16位含大小写字母+数字+符号的密码，熵值约100位，需要数十亿年才能暴力破解。': 'Based on password entropy and modern GPU clusters capable of 100 billion guesses per second. A 16-char password with all character types has ~100 bits of entropy — billions of years to brute-force.',
    '口令短语 vs 随机密码，哪个更好？': 'Passphrase vs random password — which is better?',
    '两者各有优势。口令短语（如 correct-horse-battery-staple）更易记忆，5个随机单词的熵值可达100+位，安全性极高。随机密码更紧凑但不易记忆。建议对需要手动输入的场景使用口令短语，对密码管理器自动填充的场景使用随机密码。': 'Both have advantages. Passphrases (e.g. correct-horse-battery-staple) are easier to remember — 5 random words = 100+ bits entropy. Random passwords are more compact but harder to remember. Use passphrases for manual entry, random passwords for password managers.',
    '密码强度条代表什么？': 'What does the password strength bar mean?',
    '密码强度基于熵值（信息熵）计算。弱（红色，&lt;30位）→ 中等（橙色，30-50位）→ 强（黄色，50-70位）→ 非常强（绿色，&gt;70位）。16位含所有字符类型的密码通常为"非常强"。': 'Based on entropy. Weak (red, <30 bits) → Medium (orange, 30-50) → Strong (yellow, 50-70) → Very Strong (green, >70). A 16-char password with all types is usually "Very Strong".',
    '排除相似字符有什么作用？': 'Why exclude similar characters?',
    '字符 0/O、1/l/I 在 Sans-serif 字体下难以区分。排除这些字符可避免手动输入时的混淆，提高密码的可读性和输入准确性。': 'Characters 0/O, 1/l/I are hard to distinguish in sans-serif fonts. Excluding them prevents manual entry errors.',
    '导出的密码文件安全吗？': 'Are exported password files secure?',
    '所有导出操作在浏览器本地完成，文件直接下载到您的设备，不会经过任何服务器。建议将导出的文件加密存储或导入密码管理器后删除原文件。': 'All exports happen locally in your browser. Files download directly to your device and never touch any server. We recommend encrypting exported files or importing them into a password manager, then deleting the originals.',
    '二维码分享如何工作？': 'How does QR code sharing work?',
    '点击密码旁的📱按钮，会生成包含该密码的二维码。用手机相机扫描即可获取密码，方便从电脑传输到手机。二维码完全在浏览器本地生成，密码不会上传到任何服务器。': 'Click 📱 to generate a QR code containing the password. Scan with your phone camera to transfer from desktop to mobile. QR codes are generated entirely in your browser — passwords are never uploaded.',
    '记忆难度评分是什么？': 'What is the memorability score?',
    '每个密码旁显示记忆难度：': 'Each password shows a memorability rating: ',
    '易记': 'Easy',
    '中等': 'Medium',
    '难记': 'Hard',
    '（如口令短语、纯数字PIN）': '(passphrases, numeric PINs)',
    '（字母+数字组合）': '(letters+numbers)',
    '（含特殊符号的随机字符串）。帮助您在安全性和可记忆性之间找到平衡。': '(random strings with symbols). Helps you balance security and memorability.',
    '密码强度说明': 'Strength Guide',
    '弱': 'Weak',
    '强': 'Strong',
    '非常强': 'Very Strong',
    '位熵': ' bits entropy',
    '新功能': 'New Features',
    '密码卡片打印': 'Password Card Print',
    '记忆难度': 'Memorability',
    '格式': 'Format',
    '无格式': 'Plain',
    '每4位空格': '4-digit Space',
    '每4位连字符': '4-digit Dash',
    '词库分类': 'Word Category',
    '通用': 'General',
    '科技': 'Tech',
    '自然': 'Nature',
    '食物': 'Food',
    '动物': 'Animals',
    '太空': 'Space',
    '安全评分': 'Security Score',
    '平均熵': 'Avg Entropy',
    '最短破解时间': 'Min Crack Time',
    '字符池大小': 'Char Pool Size',
    '密码已复制到剪贴板': 'Password copied to clipboard',
    '已复制全部': 'All copied',
    '已导出': 'Exported',
    '文件': 'file',
    '在线密码生成器 · 纯前端处理 · 密码不会上传到任何服务器': 'Online Password Generator · Pure Frontend · Passwords Never Uploaded',
    '反馈': 'Feedback',
    '无历史记录': 'No history yet',
    '密码': 'Password',
    '免费密码生成器 - 强随机密码 | 口令短语 | PIN码 | 实时生成': 'Free Password Generator - Strong Random Passwords | Passphrase | PIN | Real-time',
    '个)': ')',
    '最多10条': 'max 10 entries',
    '🔍': '🔍',
}

# Apply translations
en = cn
for cn_text, en_text in trans_map.items():
    en = en.replace(cn_text, en_text)

# Fix HTML lang
en = en.replace('<html lang="zh-CN">', '<html lang="en">')

# Fix canonical
en = en.replace(
    'href="https://free-toolbase.com/password-generator/"',
    'href="https://free-toolbase.com/en/password-generator/"'
)

# Fix hreflang
en = en.replace('hreflang="zh-CN"', 'hreflang="en"')

# Fix og:locale
en = en.replace('zh_CN', 'en_US')

# Fix alternates
en = re.sub(
    r'<link rel="alternate" hreflang="zh-CN" href="[^"]*"',
    '<link rel="alternate" hreflang="en" href="https://free-toolbase.com/en/password-generator/"',
    en
)

cn_pattern = re.compile(r'[\u4e00-\u9fff]')
remaining_cn = len(cn_pattern.findall(en))
orig_cn = len(cn_pattern.findall(cn))
print(f'Original CN chars: {orig_cn}')
print(f'Remaining CN chars: {remaining_cn}')

# Show remaining CN if any
if remaining_cn > 0:
    for i, line in enumerate(en.split('\n'), 1):
        matches = cn_pattern.findall(line)
        if matches:
            # Skip script/style blocks
            print(f'  Line {i}: {line.strip()[:100]}')

with open('en/password-generator/index.html', 'w') as f:
    f.write(en)
print('Written en/password-generator/index.html')